### ISSUE TITLE
[Android] Production app crashing in runtime

### ISSUE THREAD

**Issue by manijak** [NativeScript/NativeScript#3901]
I was asked by the NativeScript-team to post my issue here, so here it goes. Thanks in advance.

On February 26th we published a new version of both our Android and iOS apps to the app stores. The published apps are based on {N} v2.4. The issues we are having are only on Android.

We use Google-Analytics innside the app, so we try to log exceptions this way. But this only logs the first line of the exception, for some reason. 

Still, we see an outrageous increase in the amount of crashes logged. Since the release date and to this date, there are in total 2240 logged exceptions/crashes.

This is just a part of them. There are more, and those are all `OutOfMemoryError` exceptions (around 400 of them): http://i.imgur.com/lf7Z95h.png

The problem we are having is that (of course) we are struggling to reproduce many of these. The one we are able to reproduce is the `OutOfMemoryError`. It happens at random, just tapping around the app, opening/closing different views. Our app is image-rich, but the images are not bigger than 200kb. We also use a image-caching plugin called `nativescript-web-image-cache`.

One other exception, `IllegalStateException (@ViewPager:populate:1167) {main}` is probably caused by our own Carousel plugin. But we are still having a hard time to reproduce the issue locally, and are still looking into it. https://github.com/manijak/nativescript-carousel

We are using Genymotion as our main emulator for Android. And for devices we have several phones (Samsung , LG, Nexus) and also a high-end tablet (Nvidia Shield K1). The only exception we have managed to reproduce is still, `OutOfMemoryError`. 

Luckily, we have some users who care, so they reported the crashes, so have have full stack-traces on some of them. I can provide those if needed. 

Please, let me know if you need more information regarding the issue(s). I could also provide access to the source code if needed. 

Apps are available here:
Android: https://play.google.com/store/apps/details?id=ba.fkzeljeznicar
iOS: https://itunes.apple.com/us/app/fk-%C5%BEeljezni%C4%8Dar/id1097417103?mt=8

**Comment by NathanaelA** [comment]
Hey Ned, I have found that doing `if (global.android) { global.GC(); }` at certain key points helps a lot.  The issue that I found with OOM's is that the images fill up the Java memory side; and so there is no memory pressure in JavaScript in NS.   So basically the Java memory runs out and no GC is triggered.   ;-(       But if I manually trigger a GC on exit of the large memory screens, it so far appears to help and deal with the issue.      

I can't 100% say if this is the solution to OOM's as my next version of the app with this code hasn't been deployed to the wild; but I haven't seen a OOM internally since doing this and I have been working on this app a lot...

**Comment by NickIliev** [comment]
@manijak can you please provide the stack trace logs. Access to the source code will also be of great help to investigate what is causing the memory leak.

**Comment by Pip3r4o** [comment]
@NathanaelA here's some more info in addition to what you said. It might be helpful to others https://docs.nativescript.org/runtimes/android/advanced-topics/memory-management

**Comment by manijak** [comment]
Did not know about the memory-management article or the 'android.GC()' function. I will look more into this for sure. Thanks!

**Comment by NickIliev** [comment]
@manijak 

Here is what we have so far:
Regarding the error log from **com.tns.NativeScriptException (com.tns.Runtime.callJSMethodNative)_2.txt**
```
java.lang.RuntimeException: Unable to resume activity {ba.fkzeljeznicar/com.tns.NativeScriptActivity}: com.tns.NativeScriptException: 
Calling js method onStart failed
TypeError: Cannot read property 'getViewTreeObserver' of undefined
File: "/data/data/ba.fkzeljeznicar/files/app/tns_modules/ui/scroll-view/scroll-view.js, line: 125, column: 21
```
This one appears to be caused by the ScrollView implemented in **_views/home/home.xml_**.
To reproduce the crash open the app >> navigate to "more" >> minimaze the app >> return back to the app >> the app will throw with this error (in debug mode) or will simply exit in production.

The reason for this to happen is because the scrollview is not yet initialized on view loaded event. I am guessing that the partials loaded inside causing this behaviour as I was unable to reproduce it in local project with the same scenario, hoever the fix was pretty easy - I've made some structural changes as follows:
in your **home.xml**
```
<ScrollView loaded="viewLoaded" ios.showsVerticalScrollIndicator="false" xmlns:fkz="views/home/">
    <StackLayout>
        <fkz:partial-articles marginTop="66"/>
        <fkz:partial-matches />
        <fkz:partial-video />
        <fkz:partial-sponsor /> 
        <fkz:partial-articles-rest />
    </StackLayout>     
</ScrollView>
```
Now that the ScrollView has guarantied single parent and is our root element there were also some minor changes in **viewLoaded** as follows:
**_home.js_**
```
exports.viewLoaded = function(args){
var dashScrollView = args.object; 
// this is now the ScrollView as we are using its loaded event
// so no need of getViewById
...
}
```

**Comment by NickIliev** [comment]
@manijak regarding the OOM issue
```
java.lang.OutOfMemoryError: Failed to allocate a 2805772 byte allocation with 2547104 free bytes and 2MB until OOM
```

As suggested by @NathanaelA  and @Pip3r4o you can manually trigger garbage coolector and also take advantage of the fandroid flags set in app/package.json to trigger the clean up of memory on some time interval. Keep in mind that the global.GC() will trigger the JavaScript garbage collector only and not the JAVA one. You can combine both tactics to force the cleanup in your project.
e.g.
open your applicaiton package.json (not the root one but the one located in your **app** folder)
**_app/package.json_**
```
"android": {
	"v8Flags": "--expose_gc",
	"memoryCheckInterval": 500,
	"freeMemoryRatio": 0.50
},
```
The best explanation what this will do can be found [here](https://docs.nativescript.org/runtimes/android/advanced-topics/memory-management#using-memorycheckinterval-and-freememoryratio-parameters)

Also as @NathanaelA mentioned yu can trigger the JavaScript GC on key points in your application (for example when the user is leaving a heavy-duty image page) with
```
exports.onUnloaded = function(args){
    console.log("view unloaded");

    if (global.android) {
        global.gc();
    }
}
```

After combining both techniques and monitoring the memory usage resulted taht the application memory has been now stable and is freed on some time interval.
I've used adb dumpsys to check the memory usage.
```
adb shell dumpsys meminfo | grep ba.fkzeljeznicar.fkzeljoapp
```

**Comment by manijak** [comment]
Thank you @NickIliev . This is very helpful. Have already managed to reproduce the crash above and fixed it. Altho I am curious on how you managed to find out that the issue was located in the `home.xml` view? The stacktrace didn't have any mention of it or any of my files. 

Also thanks for the `gc` tips. Very helpful. Do the same rules apply for iOS? I have not had any issues reported from iOS users, nor anything logged in analytics.

**Comment by NickIliev** [comment]
@manijak for iOS you can trigger the GC following [these instructions ](https://github.com/NativeScript/NativeScript/issues/3887#issuecomment-290438587). Overall in iOS memory usage problems are much harder to achieve, but if you observe a high memory consumption you can call the GC with
```
global.__collect(); 
```
Once again it will be rarely needed but you can observe your app memory usage with Instruments  >> Allocations app in Mac and decide if this is needed at all.
As for the issue in `home.xml` - I've noticed that the error is thrown at [this line ](https://github.com/NativeScript/NativeScript/blob/master/tns-core-modules/ui/scroll-view/scroll-view.android.ts#L104) (the log shows the same line in the transpiled JS file but this is the corresponding one in the source TS file) so I've simply started to "turn off" all scroll views in the app until the error was no longer reproducible.

As for the other error logs - I am currently unable to reproduce them at my side - only once I got a page transition problem but it was inconsistent and does not re-appear. If you come to a stable way to hit one of those please let me know.

**Comment by manijak** [comment]
I was just testing the memory management tips you guys provided. I updated app/package.json and included the memorycheck parts. Also added two `global.GC();` calls in two different views (article & player-details). 

Started up the app, and just tapped around, article 1, article 2, match-details, tab2, tab3, tab4, then lastly player-details. As soon as I tapped back, before it managed to hit `unloaded` I get the exception below. Could this be the one you were after? I managed to get it two times in a row, but its not consistent. It needs to reach a certain memory threshold I guess...

I think issue here is that my main-view (TabView) contains a lot of complex partial-views (for each tab). 
And that is where most of the memory goes. Would it be wise to force GC when leaving the main-view?

```
*** StackTrace *** : An uncaught Exception occurred on "main" thread.
java.lang.OutOfMemoryError: Failed to allocate a 2188332 byte allocation with 1938816 free bytes and 1893KB until OOM
	at dalvik.system.VMRuntime.newNonMovableArray(Native Method)
	at android.graphics.Bitmap.nativeCreate(Native Method)
	at android.graphics.Bitmap.createBitmap(Bitmap.java:831)
	at android.graphics.Bitmap.createBitmap(Bitmap.java:808)
	at android.graphics.Bitmap.createBitmap(Bitmap.java:775)
	at android.transition.TransitionUtils.createViewBitmap(TransitionUtils.java:165)
	at android.transition.TransitionUtils.copyViewImage(TransitionUtils.java:102)
	at android.transition.Visibility.onDisappear(Visibility.java:384)
	at android.transition.Visibility.createAnimator(Visibility.java:247)
	at android.transition.Transition.createAnimators(Transition.java:732)
	at android.transition.TransitionSet.createAnimators(TransitionSet.java:405)
	at android.transition.Transition.playTransition(Transition.java:1777)
	at android.transition.TransitionManager$MultiListener.onPreDraw(TransitionManager.java:306)
	at android.view.ViewTreeObserver.dispatchOnPreDraw(ViewTreeObserver.java:944)
	at android.view.ViewRootImpl.performTraversals(ViewRootImpl.java:2055)
	at android.view.ViewRootImpl.doTraversal(ViewRootImpl.java:1107)
	at android.view.ViewRootImpl$TraversalRunnable.run(ViewRootImpl.java:6013)
	at android.view.Choreographer$CallbackRecord.run(Choreographer.java:858)
	at android.view.Choreographer.doCallbacks(Choreographer.java:670)
	at android.view.Choreographer.doFrame(Choreographer.java:606)
	at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:844)
	at android.os.Handler.handleCallback(Handler.java:739)
	at android.os.Handler.dispatchMessage(Handler.java:95)
	at android.os.Looper.loop(Looper.java:148)
	at android.app.ActivityThread.main(ActivityThread.java:5417)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)
```

And the third time I got this one. This is what concerns me, they are so random and sometimes different from one another....
```
 *** StackTrace *** : An uncaught Exception occurred on "main" thread.
  com.tns.NativeScriptException: No weak reference found. Attempt to use cleared object reference id=22458
		at com.tns.Runtime.getJavaObjectByID(Runtime.java:971)
		at com.tns.Runtime.callJSMethodNative(Native Method)
		at com.tns.Runtime.dispatchCallJSMethodNative(Runtime.java:1197)
		at com.tns.Runtime.callJSMethodImpl(Runtime.java:1061)
		at com.tns.Runtime.callJSMethod(Runtime.java:1047)
		at com.tns.Runtime.callJSMethod(Runtime.java:1028)
		at com.tns.Runtime.callJSMethod(Runtime.java:1018)
		at com.tns.gen.android.transition.Transition_TransitionListener.onTransitionEnd(android.transition.Transition$TransitionListener.java)
		at android.transition.Transition.end(Transition.java:1924)
		at android.transition.Transition$3.onAnimationEnd(Transition.java:1875)
		at android.animation.ValueAnimator.endAnimation(ValueAnimator.java:1239)
		at android.animation.ValueAnimator$AnimationHandler.doAnimationFrame(ValueAnimator.java:766)
		at android.animation.ValueAnimator$AnimationHandler$1.run(ValueAnimator.java:801)
		at android.view.Choreographer$CallbackRecord.run(Choreographer.java:858)
		at android.view.Choreographer.doCallbacks(Choreographer.java:670)
		at android.view.Choreographer.doFrame(Choreographer.java:603)
		at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:844)
		at android.os.Handler.handleCallback(Handler.java:739)
		at android.os.Handler.dispatchMessage(Handler.java:95)
		at android.os.Looper.loop(Looper.java:148)
		at android.app.ActivityThread.main(ActivityThread.java:5417)
		at java.lang.reflect.Method.invoke(Native Method)
		at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)
		at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)
```

**Comment by danielgek** [comment]
@NickIliev article link is broken

**Comment by NathanaelA** [comment]
@NickIliev - I would hope that manually calling `global.GC()` doesn't cause the `No Weak ref` error.    It is used in the TNS core modules (which is why it is required to be exposed from the package.json).   

And even w/o using global.GC() my clients are hitting the `No Weak Ref` error because of internal GC's occurring...   

@danielgek -- the link seems to be fixed now.

**Comment by NickIliev** [comment]
@manijak I've noticed that the plugin [nativescript-web-image-cache](https://github.com/VideoSpike/nativescript-web-image-cache) has a cache clearing method. Perhaps you can try to manually clear the cache (when needed) using the plugin preferable to calling the GC.

```
var imageCacheModule=require("nativescript-web-image-cache");
imageCacheModule.clearCache();
```

**Comment by manijak** [comment]
@NickIliev Yeah I've tried that before (also just now), it didn't have any major impact on memory and I received a OOM at the end. It freed up between 500-2000 kb in total. 

The biggest jumps in memory seem to be when I go back to main-view from another Page. And also when switching Tabs. 

I don't know, but I feel that calling `clearCache()` too often would defeat the purpose of it.

**Comment by manijak** [comment]
@NickIliev Any more tips you could recommend regarding the issues in my app? I haven't managed to improve the memory usage. It seems to increase the more views one visits.  I've tried with GC settings, and using the app normally, still it ends with OOM. Also tried forcing `clearCache` on the image plugin, but the decrease in memory is very small. 

Thanks again.

**Comment by NickIliev** [comment]
Hey, @manijak we've managed to stably reproduce the no Weak Reference related issue and @Pip3r4o is currently looking at it. I will let you know if we have further update on that.

Meanwhile some key notes regarding how to improve your application performance and also eliminate the out-of-memory issues.(without the need to use GCs)

- I've noticed that you are using Repeater in your partial-views (e.g. **_partial-articles-rest.xml_** and **_partial-matches.xm_l** and maybe on other places as well) and those partials are nested in your heaviest layout (**_home.xml_**) which is loaded in the first tab of your entry point file. Now notice that the main differences within ListView and Repeater are that the ListView is optimized and is using virtualisation and recycling of cells. **The Repeater lacks both!**. This said it is always good to use Repeater only for a small number of items when it will be expected for all items to be visible. For complex scenarios (like the home page of your application) the usage of Repeaters is the main reason for the increasing memory usage and out of memory problems.
As a rule of thumb, I would suggest to always use ListView when creating heavy duty galleries like pages with a lot of dynamic content which will use images and data..
[ This article ](https://medium.com/@alexander.vakrilov/faster-nativescript-listview-with-multiple-item-templates-8f903a32e48f)explains why it is important to use ListView (in the context of Angular application but the same principle is 100 percent valid in NativeScript core application) and also explains how virtualization and recycling are helping to improve greatly the overall performance while decreasing the memory consumption. 
**_Suggested solution_**: I would suggest recreating your partials (and other pages where Repeater is used) with ListView.

- Another technique I would recommend is to minimize the nesting of the layout as much as possible. The more complex your nesting is the more resources your app will need and in the mobile world where some phones are still with limited hardware resources, this is important issue.
Example from the code in your application:
**main-view.xml**
```
<StackLayout class="spinner-container" id="spinner">
    <StackLayout height="100%" widht="100%">
        <!-- some more compomnents follows here-->
    </StackLayout>
</StackLayout>
```
can reduce the nesting and in the same time without chaning the actual UX
```
    <StackLayout class="spinner-container" id="spinner" height="100%" widht="100%">
        <!-- some more compomnents follows here-->
    </StackLayout>
```

In the same page, you have also a GridLayout which contains the TabView which again can be reduced just to a TabView. Some pages like **partial-view.xml** can reduce the nested layouts greatly!
The less nesting you have the fewer operations will be executed when your UI is recreated (e.g. like propagating the inherited CSS) and the faster your app will load while using minimum resources.

Some of the nested layouts are kinda hidden by the custom components (in your case partials).
For example the page **home.xml** looks like this
```
<ScrollView>
   <StackLayout>
        <partialz-one />
        <partialz-two/>
        <partialz-three/>
   </StackLayout>
</ScrollView>
```
But when opening one of the partial we see another empty  stacklayout as a container.
**partial-sponsor.xml**
```

<StackLayout>
    <CardView />
</StackLayout>
```

Again the nested layout can be eliminated without sacrificing nothing from UX perspective.
The rule of thumb should be to use as less as possible nesting - even [the most complex layouts](http://developer.telerik.com/featured/demystifying-nativescript-layouts/) can be recreated with only a dew levels of nesting

- the third thing that I've noticed is that from the main page >> tapping on a news from the first tab  will direct you to the detailed news page. Going backward will trigger the whole application to forcibly fetch the whole online information again (using a method called **refreshAllViews** in **pageLoaded** in **_main-view.js_** ). Is that really necessary when the user is only returning back to the news page **tab**? In the best case scenario, I would expect that returning from news details to the main news page to refresh only the news section (and not the videos, fixtures and teams). As this is a very heavy duty operation it causes a delay when navigating back

**Comment by manijak** [comment]
Thank you so much @NickIliev! This is what I was after. 
I wasn't sure that some of my views were unnecessary complex (structure-wise). But it's good to know, I'll be doing some cleanup this Easter :smile: 

I had been testing with the ListView instead of Repeater in some places (not all, I admit). But had issues with height since the parent is most of the time a ScrollView (when mixing components, like `home.xml`). Will have to figure out something here. 👍 

Props for finding out the Weak Reference Error! 
The third issue you found is actually a bug! Should definitely not do that. 🐛 

I'll be doing some refactoring in the coming days, will report back the results. 
Thanks again for your time. You guys are the best 🥇

**Comment by manijak** [comment]
@NickIliev , regarding the "nested-views" issue:

- TabView must be nested (on Android) if one is using overlays, like ActivityIndicator. Otherwise the overlay appears behind all elements inside each TabViewItem. Will post a separate issue on this with a demo app. 

- When it comes to the partial-views in my `home.xml` (and others). As you can see from the example you posted above, it also, needs to be nested because of the xmlns namespace declarations. If there is another way to do this, please do let me know.
```
<StackLayout xmlns:ned="nativescript-carousel" xmlns:ic="nativescript-web-image-cache">
    <ned:CardView />
</StackLayout>
```
This part is a bit anoying, since we need to declare all namespaces in each view, partial or not. Would be great if partial views could inherit the xmlns namespace this from the parent Page-view.

**Comment by manijak** [comment]
@NickIliev @Pip3r4o 
I've been doing some cleanup in my app, as much as I was allowed to due to the facts posted in my previous comment. While testing I haven't seen any OOM so far. 

But. After I delt with this: https://github.com/NativeScript/NativeScript/issues/4004
I keep getting the `No weak reference found` error more frequently. I have managed to reproduce it every time now both on emulator and device. Could this be related to the SDK in some way?

It happens when navigating `back` from a Page to the main-view. 
```
4-15 07:19:00.154  1881  1881 W System.err:    at com.tns.Runtime.callJSMethodNative(Native Method)
04-15 07:19:00.154  1881  1881 W System.err:    at com.tns.Runtime.dispatchCallJSMethodNative(Runtime.java:1197)
04-15 07:19:00.154  1881  1881 W System.err:    at com.tns.Runtime.callJSMethodImpl(Runtime.java:1061)
04-15 07:19:00.154  1881  1881 W System.err:    at com.tns.Runtime.callJSMethod(Runtime.java:1047)
04-15 07:19:00.154  1881  1881 W System.err:    at com.tns.Runtime.callJSMethod(Runtime.java:1028)
04-15 07:19:00.154  1881  1881 W System.err:    at com.tns.Runtime.callJSMethod(Runtime.java:1018)
JS:  *** StackTrace *** : An uncaught Exception occurred on "main" thread.
JS: com.tns.NativeScriptException: No weak reference found. Attempt to use cleared object reference id=20925
JS:     at com.tns.Runtime.getJavaObjectByID(Runtime.java:971)
JS:     at com.tns.Runtime.callJSMethodNative(Native Method)
JS:     at com.tns.Runtime.dispatchCallJSMethodNative(Runtime.java:1197)
JS:     at com.tns.Runtime.callJSMethodImpl(Runtime.java:1061)
JS:     at com.tns.Runtime.callJSMethod(Runtime.java:1047)
JS:     at com.tns.Runtime.callJSMethod(Runtime.java:1028)
JS:     at com.tns.Runtime.callJSMethod(Runtime.java:1018)
JS:     at com.tns.gen.android.transition.Transition_TransitionListener.onTransitionEnd(android.transition.Transition$TransitionListene
r.java)
JS:     at android.transition.Transition.end(Transition.java:1924)
JS:     at android.transition.Transition$3.onAnimationEnd(Transition.java:1875)
JS:     at android.animation.ValueAnimator.endAnimation(ValueAnimator.java:1153)
JS:     at android.animation.ValueAnimator.doAnimationFrame(ValueAnimator.java:1313)
JS:     at android.animation.AnimationHandler.doAnimationFrame(AnimationHandler.java:146)
JS:     at android.animation.AnimationHandler.-wrap2(AnimationHandler.java)
JS:     at android.animation.AnimationHandler$1.doFrame(AnimationHandler.java:54)
JS:     at android.view.Choreographer$CallbackRecord.run(Choreographer.java:872)
JS:     at android.view.Choreographer.doCallbacks(Choreographer.java:686)
JS:     at android.view.Choreographer.doFrame(Choreographer.java:618)
JS:     at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:860)
JS:     at android.os.Handler.handleCallback(Handler.java:751)
JS:     at android.os.Handler.dispatchMessage(Handler.java:95)
JS:     at android.os.Looper.loop(Looper.java:154)
JS:     at android.app.ActivityThread.main(ActivityThread.java:6119)
JS:     at java.lang.reflect.Method.invoke(Native Method)
JS:     at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:886)
JS:     at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:776)
```

**Comment by Pip3r4o** [comment]
@manijak I have isolated the issue and can reliably reproduce it. I have also traced the cause of it, and am going to start working on it as soon as the holidays are over.

**Comment by Pip3r4o** [comment]
@manijak I have made a PR(https://github.com/NativeScript/android-runtime/pull/744) that addressed part of the causes for the `No weak reference found` exceptions. Unfortunately I observed this faulty behavior when the Repeater view was used in your project. Upon navigating back from the `articles-page` for example the crash is easy to reproduce. I believe it has something to do with the dependency properties used to store a reference to the parent view (stack-layout by default) getting out of scope at some point without strong reference to the Repeater javascript instance. I need a little more time to look into it. 

Meanwhile, consider avoiding using the Repeater view, and if it's inevitable - inflate layouts manually from code.

**Comment by manijak** [comment]
@Pip3r4o Thanks for the feedback. 
Yeah you are right that both the `main-view` and the `article-page` do have a repeater each. If I understood your correctly, we should move away from `Repeater`, and just add the layouts (views) via code?

**Comment by Pip3r4o** [comment]
@manijak as I mentioned earlier - Repeater may no longer be a problem when you upgrade your application to use the 3.0.0 bits. If upgrading everything to 3.0.0 is too big a step right now, you could use just the 3.0.0 runtime (it's completely backwards compatible with 2.5 modules and CLI), which will solve the main GC problem. For the time being you could also remove/rewrite the views to not use a repeater.

**Comment by manijak** [comment]
Hi again, @Pip3r4o & @NickIliev.
I recently updated TNS and all my apps to 3.0.1. I was hoping that the issue with `no weak reference found` would be solved in this mayor release, but that is not the case it seems. 

The issue still seems to resolve around transitions and animations... I already got this three times when navigating back to main view. 

```
05-27 07:14:15.345  2468  2468 W System.err:    at com.tns.Runtime.callJSMethodNative(Native Method)
05-27 07:14:15.345  2468  2468 W System.err:    at com.tns.Runtime.dispatchCallJSMethodNative(Runtime.java:1021)
05-27 07:14:15.345  2468  2468 W System.err:    at com.tns.Runtime.callJSMethodImpl(Runtime.java:903)
05-27 07:14:15.345  2468  2468 W System.err:    at com.tns.Runtime.callJSMethod(Runtime.java:890)
05-27 07:14:15.345  2468  2468 W System.err:    at com.tns.Runtime.callJSMethod(Runtime.java:874)
05-27 07:14:15.345  2468  2468 W System.err:    at com.tns.Runtime.callJSMethod(Runtime.java:866)
JS:  *** StackTrace *** : An uncaught Exception occurred on "main" thread.
JS: com.tns.NativeScriptException: No weak reference found. Attempt to use cleared object reference id=20194
JS:     at com.tns.Runtime.getJavaObjectByID(Runtime.java:825)
JS:     at com.tns.Runtime.callJSMethodNative(Native Method)
JS:     at com.tns.Runtime.dispatchCallJSMethodNative(Runtime.java:1021)
JS:     at com.tns.Runtime.callJSMethodImpl(Runtime.java:903)
JS:     at com.tns.Runtime.callJSMethod(Runtime.java:890)
JS:     at com.tns.Runtime.callJSMethod(Runtime.java:874)
JS:     at com.tns.Runtime.callJSMethod(Runtime.java:866)
JS:     at com.tns.gen.android.transition.Transition_TransitionListener.onTransitionEnd(Transition_TransitionListener.java:17)
JS:     at android.transition.Transition.end(Transition.java:1924)
JS:     at android.transition.Transition$3.onAnimationEnd(Transition.java:1875)
JS:     at android.animation.ValueAnimator.endAnimation(ValueAnimator.java:1149)
JS:     at android.animation.ValueAnimator.doAnimationFrame(ValueAnimator.java:1309)
JS:     at android.animation.AnimationHandler.doAnimationFrame(AnimationHandler.java:146)
JS:     at android.animation.AnimationHandler.-wrap2(AnimationHandler.java)
JS:     at android.animation.AnimationHandler$1.doFrame(AnimationHandler.java:54)
JS:     at android.view.Choreographer$CallbackRecord.run(Choreographer.java:869)
JS:     at android.view.Choreographer.doCallbacks(Choreographer.java:683)
JS:     at android.view.Choreographer.doFrame(Choreographer.java:616)
JS:     at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:857)
JS:     at android.os.Handler.handleCallback(Handler.java:751)
JS:     at android.os.Handler.dispatchMessage(Handler.java:95)
JS:     at android.os.Looper.loop(Looper.java:154)
JS:     at android.app.ActivityThread.main(ActivityThread.java:6077)
JS:     at java.lang.reflect.Method.invoke(Native Method)
JS:     at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:865)
JS:     at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:755)
```

**Comment by Pip3r4o** [comment]
@manijak at this point I can only suggest getting rid of all the repeaters, though our manual testing didn't bring the bug to light after upgrading a buggy project to all components 3.0

**Comment by manijak** [comment]
The main view has only one Repeater, with 4 items, and is used to display items horizontally. There is no ListView that would display items this way...

**Comment by NathanaelA** [comment]
@Pip3r4o - none of my views have any repeaters; and I'm seeing this in 2.5 during navigations.    Does your fix included in 3.0 fix the navigation issue, or just repeaters?

**Comment by Pip3r4o** [comment]
@NathanaelA the fix merely amplified some of the GC logic, it does not look at navigation, nor repeaters in particular. During the troubleshooting of manijak's app it became clear that the repeater was a somewhat problematic view, it also pointed me in the direction of the case that was initially omitted during GC. The changes in the android runtime combined with the new tns-core-modules property system appeared to have eliminated the crashes from the isolated cases that would otherwise crash with the 2.5 bits.

**Comment by manijak** [comment]
I was just about to ask for an update on this issue, and decided to run one more test in my app. 
Every time I got this `no weak reference found` error, it was when navigating back to the main view. 

So, I removed the only partial-view from main that contained a `Repeater`. Keep in mind this is a simple Repeater that generated max 4 items that scroll horizontally. Then, ran the same test (navigating to an `article-view` and back to main), the app still crashed with the above exception. After like, 2-3 tries. 

Then I realised that my `article-view` also contained a `Repeater`. Now, this Repeater is used when the article-data had a youtube-url. Which is very rare. We show in all around 15 articles on the main-view, and perhaps one of them would have a youtube-url, tops two.  

Here is the snippet:
```
...
<Repeater items="{{ video }}">
    <Repeater.itemTemplate>
        <WebView height="300" margin="0,0,40,0" url="{{ 'https://www.youtube.com/embed/' + video }}" />
    </Repeater.itemTemplate>
</Repeater>
...
```
After removing this code. The app stopped crashing... No exceptions, got to tap around several times, before it OOM-ed (different issue). 

I added back the partial-view on main that had the only Repeater. Tested several times, and it still ran fine. But when I added back the above code (in the article-view), it started crashing again... 

Now. The strange part (for me) is why would this Repeater be an issue, and not the one on the main-view? It almost never has any content, the `video` property it is bound to, is `null` most of the time. Could that be the issue here?

**Comment by Pip3r4o** [comment]
@manijak Thanks for your elaborate report on the findings! We aren't sure yet why the Repeater causes so many problems, I only knew that it was the Repeater all along, even if the bound property was empty, or when it contained few elements.

As an alternative to the Repeater component, which I believe you use only to list items horizontally - you could use the nativescript-telerik-ui's RadListView whose orientation can be set to suit your needs. I understand that will increase the application size slightly, but it might be worth a shot!

**Comment by NathanaelA** [comment]
@manijak - Please note; be very very careful using the RadListView; it has lots of bugs and not all of them can be worked around (no source).   It works in some cases; but in my experience, if I can avoid using a RLV (especially on iOS) I will, in the cases I have to use it; I inform the end user that the extra features will bring random display glitches (especially on rotation) and can cause random crashing...   (Most of these bugs have been reported to Telerik already...  None have been verified fixed yet  :-(   )

**Comment by ginev** [comment]
@NathanaelA - please allow me to disagree with you on that.
The link that you have provided lists issues on the NativeScript UI's public repository which people use for various reasons, many of which are simply handicaps that they have hit while developing but which were later resolved. They were not even bugs with the component. Other people report stuff and stop sharing information after some time, especially when it comes to reproducing a problem they report. They might simply have resolved it. We do not close all off the issues as they might be helpful for people having similar cases. It is, to say the least, pretty superficial to judge based on that. Moreover, some of the issues you see there are simply logged by us to inform people what we're working/planning to work on for the near future. On top of that, many of the issues in the Feedback repository were migrated from the SDK Samples repositories by ourselves when we decided that we will stop using them for feedback from customers. They contain useful information and that is why we decided not to close them.

As for your experience with using RadListView - we have a couple of statements to make:
- we have thoroughly reviewed your reports and have addressed issues that we were able to reproduce
- in our internal issue tracking system there are a bunch of issues that are addressed and closed during the last several months
- we are taking RadListView pretty seriously as the backbone of every mobile app and thus are treating issues related to it with highest priority

For example, our next major effort is enabling Grouping, Filtering and Sorting as it was previously postponed due to various reasons.

If you think there are still pending issues with RadListView that make it difficult to use, please let us know by providing enough information for us to reproduce.

**Comment by NathanaelA** [comment]
@ginev - You have to present a positive view of TUI, I fully understand that.   But my experience does not match your statements, I can only hope you have been provided more resources so that my experiences with RLV will change for the positive over time.    I'll even preface this that I have NOT tried the latest version, so it might be vastly better, since my last `fun` attempt a couple weeks ago.

Don't you think the `correct` thing to do is let Ned have a heads up that he `might` be walking into a different hole with this solution?    I know I personally would be appreciative if someone warned me plugin x might cause y.      

The single link I provided was so that he could get an idea of some of the `possible` issues associated with RLV.   I just pointed to the repo so `HE` can look at it, and I even was being nice and diplomatic and said `might` cause him issues.     I was trying to let HIM make an informed decision, based on what he could read without any of my **many** negative experiences with RLV prejudicing the issue any more beyond the simple warning "it has its own issues".       

I have a lot of data to refute what you stated above, but honestly the only thing I was trying to do here was let Ned know that it is not a silver bullet and that RLV has some issues that `MIGHT` cause him issues, so that he is aware of it.   If you wish to make it into a larger issue; that is your prerogative.

**Comment by NathanaelA** [comment]
@Pip3r4o - To put this back on track; my app does NOT have any `repeaters` and I see the `no weak reference` error frequently, the callstacks seems to always lead to the transition.  I will be updating the app to the 3.x runtimes to see if it helps; but last thing I read from you was that the fix you provided doesn't deal with transitions. 

 So this issue may still be a serious issue for Android has there been any work on getting detail logging so that we can trace this down.   I need some way to get access to the detailed log to send via my bug reporting.    I plan on doing a release later this week hopefully.

**Comment by manijak** [comment]
I also think it has something to do with transitions or animations. I just re-added all my Repeaters, but removed animated transitions when navigating to other views. 

The app did not crash once, it didn't even OOM. App also ran smoother, faster. But it looks awful when switching views... Very interesting. 

Regarding the RLV, thanks for the heads up. I own the pro/licensed version of nativescript-ui and it just doesn't suit my needs for now. Not until each component is split into its own plugin. I do not want to bloat my app needlessly.

**Comment by ginev** [comment]
@manijak - separating the components sounds reasonable. The first step would probably be to separate the List and SideDrawer as they are the most popular ones.

**Comment by PanayotCankov** [comment]
This looks like a transition listener that got prematurely collected:
`com.tns.gen.android.transition.Transition_TransitionListener.onTransitionEnd`

**Comment by hshristov** [comment]
@manijak Is the issue with
```
 No weak reference found. Attempt to use cleared object reference id=20194
JS:     at com.tns.Runtime.getJavaObjectByID(Runtime.java:825)
...
com.tns.gen.android.transition.Transition_TransitionListener.onTransitionEnd(Transition_TransitionListener.java:17)
JS:     at android.transition.Transition.end(Transition.java:1924)
```

still reproducible?

**Comment by manijak** [comment]
@hshristov To be honest, our app is still on version 2.4 of TNS. We are planning to release a new version around the holidays that will be based on 3.4. Will report back on the outcome.

**Comment by Tronix117** [comment]
I can reproduce it, the key is to navigate quickly between screens, I use navigation other than `fade` though, maybe the `fade` transition "hides" the issue

**Comment by NickIliev** [comment]
Closing due to inactivity
